src/ui/pages/SendForm/shared/prepareSendData.ts

```diff
@@ -0,0 +1,249 @@
+    tokenAssetCode,
+    gasLimit,
+    nftAmount,
+    nftId,
+  } = formState;
+  if (!from || !to || !tokenChain) {
+    return EMPTY_SEND_DATA;
+  }
+  if (isEthereumAddress(from)) {
+    const networksStore = await getNetworksStore();
+    const network = await networksStore.fetchNetworkById(tokenChain);
+    const chain = createChain(network.id);
+    const chainId = Networks.getChainId(network);
+
+    let tx: IncomingTransaction;
+    if (type === 'nft') {
+      if (!nftAmount || !nftId) {
+        return EMPTY_SEND_DATA;
+      }
+      const nftPosition = await getNftPosition(client, from, formState);
+      tx = createSendNFTTransaction({
+        chainId,
+        from,
+        to,
+        nft: nftPosition,
+        amount: nftAmount,
+      });
+    } else {
+      if (!tokenAssetCode || !tokenValue) {
+        return EMPTY_SEND_DATA;
+      }
+      invariant(
+        position?.asset.asset_code === tokenAssetCode,
+        'Position must mach formState.tokenAssetCode'
+      );
+      invariant(
+        getAssetImplementationInChain({ asset: position.asset, chain }),
+        'Asset must exist on chain'
+      );
+      const tokenAddressInChain = getAddress({ asset: position.asset, chain });
+      if (tokenAddressInChain === undefined) {
+        throw new Error('Token implementation is unknown in selected chain');
+      }
+      const isNativeAsset = Networks.isNativeAsset(position.asset, network);
+      tx = createSendNativeOrContractTransaction({
+        chainId,
+        from,
+        to,
+        inputToken: tokenAddressInChain,
+        tokenInterface: isNativeAsset ? 'native' : 'erc20',
+        value: commonToBase(
+          tokenValue,
+          getDecimals({ asset: position.asset, chain })
+        ).toFixed(),
+      });
+    }
+    if (gasLimit) {
+      invariant(isNumeric(gasLimit), 'Gas limit must be numeric');
+      tx.gasLimit = valueToHex(gasLimit);
+      tx.gas = valueToHex(gasLimit);
+    } else {
+      const gas = await estimateGasForNetwork(tx, network);
+      tx.gasLimit = valueToHex(gas);
+      tx.gas = valueToHex(gas);
+    }
+    let nonce = formState.nonce;
+    if (network.supports_sponsored_transactions && nonce == null) {
+      const networks = await networksStore.load();
+      const { value: latestNonce } = await uiGetBestKnownTransactionCount({
+        address: from,
+        chain,
+        networks,
+        defaultBlock: 'pending',
+      });
+      nonce = String(latestNonce);
+    }
+    tx = await applyConfigurationAsync({
+      chain,
+      formState: { ...formState, nonce },
+      transaction: tx,
+    });
+    let eligibility: SendSubmitData['paymasterEligibility'] = null;
+    if (network.supports_sponsored_transactions) {
+      eligibility = await getEligibility(tx);
+      if (eligibility.data.eligible) {
+        tx = await getPaymasterTx(tx);
+      }
+    }
+    assertProp(tx, 'chainId');
+    assertProp(tx, 'from');
+    return {
+      networkId: chain,
```

Comment:
is it ok that networkId has a Chain type?

Reply:
I'm considering this to be a part of future refactoring of `Chain` :)